### DISCUSSION TITLE
Could it be a problem to use a fetch?

### DISCUSSION THREAD

**Discussion by krutoo** [nodejs/undici#2979] [Q&A] [answered]
Hi, in readme of this project I see this chapter:
https://github.com/nodejs/undici?tab=readme-ov-file#garbage-collection

This section says that the body **must be read**.

I have small library that makes possible to add "middleware" for `fetch` function like this:
```js
async function middleware (request, next) {
  doSomethingBefore();

  const response = await next(request);

  doSomethingAfter();

  return response;
}
```

Theoretically, I can write an middleware that adds some kind of validation of the response based on headers for example and not on body like this:

```js
async function middleware (request, next) {
  const response = await next(request);

  if (isInvalidHeaders(response)) {
    throw Error('Validation error...')
  }

  return response;
}
```

In this case, the **body will not be read** since the promise will go into the catch state.

In real world applications there can be a lot of requests that would not pass such validations

Please explain whether this approach can cause problems with memory consumption?

**Comment by ronag** [comment, marked as answer]
It will be a problem. You can dump the body before throwing.

```js
async function middleware (request, next) {
  const response = await next(request);

  if (isInvalidHeaders(response)) {
    dump(response)
    throw Error('Validation error...')
  }

  return response;
}
async function dump(response) {
  try {
    for await (const chunk of response) {
      return
    }
  } catch { 
    // Do nothing
  }
} 
```

**Reply by krutoo** [reply]
@ronag Thanks a lot. Can I use this solution in browser? 

Can I just use response.body.cancel() instead?

**Reply by ronag** [reply]
I think cancel should be fine as well

**Reply by ronag** [reply]
Just be careful of any async errors

**Reply by krutoo** [reply]
> Just be careful of any async errors

@ronag Can you give some details about it?

**Reply by ronag** [reply]
I think cancel returns a promise that can throw?

**Reply by krutoo** [reply]
@ronag exactly, thank you

**Reply by jdalton** [reply]
Would this be `response.body?.dump()`?

**Reply by krutoo** [reply]
> Would this be `response.body?.dump()`?

@jdalton is it out of Fetch API specification, just for Node.js?
